src/libraries/System.Private.CoreLib/src/System/Threading/ExecutionContext.cs

```diff
@@ -88,9 +88,12 @@ public void GetObjectData(SerializationInfo info, StreamingContext context)
 
             if (m_localValues == null || AsyncLocalValueMap.IsEmpty(m_localValues))
             {
-                return isFlowSuppressed ?
-                    DefaultFlowSuppressed :
-                    null; // implies the default context
+                if (!isFlowSuppressed)
+                    return null; // implies the default context
+
+                s_DefaultFlowSuppressed ??= new ExecutionContext(AsyncLocalValueMap.Empty, Array.Empty<IAsyncLocal>(), isFlowSuppressed: true);
+                Thread.MemoryBarrier();
```

Comment:
Why MemoryBarrier and not just volatile ?

Reply:
(MemoryBarrier is full acquire + release memory barrier. It is usually more expensive than just volatile.)

Reply:
Yeah, we should make s_defaultFlowSuppressed be volatile, and then this can just be:
```C#
return s_defaultFlowSuppressed ??= new ExecutionContext(AsyncLocalValueMap.Empty, Array.Empty<IAsyncLocal>(), isFlowSuppressed: true);
```

Reply:
> Why MemoryBarrier and not just volatile ?

I'm not a fan of "hidden" fences semantic and it would also generate incorrect code for this case. The volatile generates only half-fences so it doesn’t prevent a write followed by a read from being swapped. 

> then this can just be:

I didn't realize a version with a temporary store is what is preferred, will update the PR.